### PR TITLE
Rsp check max_spdm_msg_size size to avoid assert

### PR DESCRIPTION
**Pull request by jyao1**
Fix https://github.com/DMTF/libspdm/issues/2207

### CODE REVIEW

File: library/spdm_responder_lib/libspdm_rsp_certificate.c [hunk 1]
```diff
@@ -111,6 +113,36 @@ libspdm_return_t libspdm_get_response_certificate(libspdm_context_t *spdm_contex
         length = (uint16_t)(spdm_context->local_context.local_cert_chain_provision_size[slot_id] -
                             offset);
     }
+
+    /* In chunk case, ensure the length request can be served, otherwise return error. */
+    if (libspdm_is_capabilities_flag_supported(spdm_context, false,
```
Comment (steven-bellock) [inline library/spdm_responder_lib/libspdm_rsp_certificate.c:118]: Why is this needed? Wouldn't `libspdm_build_response` handle chunking like for all the other messages?

Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_certificate.c:118]: The length must be calculated and checked here properly.

Otherwise, it will cause ASSERT at line 152 - `LIBSPDM_ASSERT(*response_size >= sizeof(spdm_certificate_response_t) + length);`

Comment (steven-bellock) [reply library/spdm_responder_lib/libspdm_rsp_certificate.c:118]: Ah so this is a case where the Integrator has provisioned a certificate chain that is larger than its `MaxSPDMmsgSize`. In that case this is true even if chunking is not enabled and the
```
LIBSPDM_ASSERT (length + sizeof(spdm_certificate_response_t) <=
                        spdm_context->local_context.capability.max_spdm_msg_size);
```
is unconditional, assuming that `max_spdm_msg_size != 0`.

Comment (steven-bellock) [reply library/spdm_responder_lib/libspdm_rsp_certificate.c:118]: That check can also be added to `libspdm_check_context` to help them catch that sooner.

File: library/spdm_responder_lib/libspdm_rsp_certificate.c [hunk 2]
```diff
@@ -111,6 +113,36 @@ libspdm_return_t libspdm_get_response_certificate(libspdm_context_t *spdm_contex
...
+            LIBSPDM_DEBUG((LIBSPDM_DEBUG_ERROR, "cert size > req max_spdm_msg_size\n"));
+            actual_size = length + sizeof(spdm_certificate_response_t);
+            return libspdm_generate_extended_error_response(
+                spdm_context, SPDM_ERROR_CODE_RESPONSE_TOO_LARGE, 0,
+                sizeof(uint32_t), (uint8_t *)&actual_size, response_size, response);
+        }
+    }
+
+    /* Adjust the length according to max_spdm_msg_size to ensure it can be fit into response buffer. */
+    if (spdm_context->connection_info.capability.max_spdm_msg_size != 0) {
```
Comment (steven-bellock) [inline library/spdm_responder_lib/libspdm_rsp_certificate.c:135]: The specification is explicit that
>If the size of a response message exceeds the size of the MaxSPDMmsgSize of the requesting SPDM endpoint, the responding SPDM endpoint shall respond with an ERROR message of ErrorCode=ResponseTooLarge.

so the Responder shouldn't truncate the response based on the Requester's `MaxSPDMmsgSize`.

File: library/spdm_responder_lib/libspdm_rsp_certificate.c (hunk 1 above)
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_certificate.c:118]: For local_context, I dont believe we need 0 check.

Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_certificate.c:118]: Uh, I am not sure if `libspdm_check_context` can help here, because the certificate chain may be provisioned later.
What we can do is only "if cert_chain != NULL, then check".

But we really dont know when `libspdm_check_context` will be invoked by the integrator. Maybe never.
Then we still need ASSERT here, even we add to `libspdm_check_context`

Comment (steven-bellock) [reply library/spdm_responder_lib/libspdm_rsp_certificate.c:118]: The assertion would remain here but it can also be checked in `libspdm_check_context`.

File: library/spdm_responder_lib/libspdm_rsp_certificate.c (hunk 2 above)
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_certificate.c]: done. This check is removed.

File: library/spdm_responder_lib/libspdm_rsp_certificate.c (hunk 1 above)
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_certificate.c:117]: agree. Done.

File: library/spdm_common_lib/libspdm_com_context_data.c [hunk 3]
```diff
@@ -1016,6 +1017,24 @@ bool libspdm_check_context (void *spdm_context)
...
+        for (index = 0; index < SPDM_MAX_SLOT_COUNT; index++) {
+            if ((context->local_context.local_cert_chain_provision_size[index] != 0) &&
+                (context->local_context.local_cert_chain_provision_size[index] +
+                 sizeof(spdm_certificate_response_t) >
+                 context->local_context.capability.max_spdm_msg_size)) {
+                LIBSPDM_DEBUG((LIBSPDM_DEBUG_ERROR,
+                               "max_spdm_msg_size (%d) must be greater than or "
+                               "equal to local_cert_chain_provision_size(%d) (%d).\n",
+                               context->local_context.capability.max_spdm_msg_size, index,
+                               context->local_context.local_cert_chain_provision_size[index]));
```
Comment (steven-bellock) [inline library/spdm_common_lib/libspdm_com_context_data.c:1033]: Needs to return `false`.

Comment (jyao1) [reply library/spdm_common_lib/libspdm_com_context_data.c:1033]: fixed.
